modules/assim.sequential/R/sda.enkf.R

```diff
@@ -129,176 +145,382 @@ sda.enkf <- function(settings,IC,prior,obs){
+    Pf   = cov(X)
+    Y    = t(obs.mean[t,])#obs$mean[t]
+    R    = diag(as.numeric(obs.sd[t,])^2)#obs$sd[t]^2
+    H    = diag(ncol(obs.mean))
+    if(processvar == FALSE){
+      K    = Pf%*%t(H)%*%solve(R+H%*%Pf%*%t(H))
+      mu.a = mu.f + K%*%(Y-H%*%mu.f)
+      Pa   = (diag(ncol(X)) - K%*%H)%*%Pf
+    } else { 
+        
+        #### initial conditions
+        bqq[1] <- length(mu.f)
+        aqq[1,,] <- diag(length(mu.f))*bqq[1]
+        
+        ### analysis of model and data
+        update = list(Y=Y, muf=mu.f, pf=solve(Pf,tol=0), aq=aqq[t,,], bq=bqq[t], r=solve(R))
+        mod <- jags.model(file=textConnection(AnalysisFilterQ),
+                          data=update,
+                          n.adapt=1000,n.chains=3,
+                          init=list(X.mod=as.vector(mu.f))) #inits for q?
+        jdat <- coda.samples(mod,variable.names=c("X","q"),n.iter=10000) 
+        
+        ## update parameters  
+        dat = as.matrix(jdat)
+        dat = dat[3000:10000,]
+        iq = grep("q",colnames(dat))
+        iX = grep("X[",colnames(dat),fixed=TRUE)
+        mu.a  = colMeans(dat[,iX])
+        Pa  = cov(dat[,iX])
+        Pa[is.na(Pa)]<- 0 
+        
+        CI.X1[,t] = quantile(dat[,iX[1]],c(0.025,0.5,0.975))
+        CI.X2[,t] = quantile(dat[,iX[2]],c(0.025,0.5,0.975))
+        
+        mq = dat[,iq] #Omega, Precision
+        q.bar = matrix(apply(mq,2,mean),length(mu.f),length(mu.f)) #Mean Omega, Precision
+        
+        col = matrix(1:length(mu.f)^2,length(mu.f),length(mu.f))
+        WV = matrix(0,length(mu.f),length(mu.f))
+        for(i in 1:length(mu.f)){
+          for(j in 1:length(mu.f)){
+            WV[i,j] <- wish.df(q.bar, X = mq, i=i, j=j, col=col[i,j])
+          }
+        }
+        
+        n = mean(WV) #n + 1
+        if(n < length(mu.f)) n = length(mu.f)
+        V = solve(q.bar)*n
+        
+        #ifelse(eigen(V)$values>0,eigen(V)$values,print("matrix not positive definite"))
+        
+        aqq[t+1,,] = V
+        bqq[t+1] = n
+    }
+
+    enkf.params[[t]] = list(mu.f = mu.f, Pf = Pf, mu.a = mu.a, Pa = Pa) 
  
  ## update state matrix
-analysis = as.data.frame(rmvnorm(nens,mu.a,Pa,method="svd"))
-names(analysis) = names(X)
- # EAKF
- if(FALSE){
-   analysis = X
-   
-   ## Math from Anderson 2001. gives correct mean but incorrect var
-   A.svd = svd(Pf)
-   F = A.svd$v
-   G = diag(sqrt(A.svd$d))
-   B.svd = svd(t(G)%*%t(F)%*%t(H)%*%solve(R)%*%H%*%F%*%G)
-   U = B.svd$v
-   B = diag((1+B.svd$d)^(-0.5))
-   A = solve(t(F))%*%t(G)*solve(t(U))%*%t(B)%*%solve(t(G))%*%t(F)
-   for(i in 1:nens){
-     analysis[i,] = t(A)%*%matrix(as.numeric(X[i,])-mu.f)+mu.a
-   }
-   
-   ## HACK IGNORNING COVARIANCE
-   for(i in 1:nens){
-     analysis[i,] = mu.a + (matrix(as.numeric(X[i,]))-mu.f)*sqrt(diag(Pa)/diag(Pf))
-   }   
-   
- }
+    analysis = as.data.frame(rmvnorm(nens,mu.a,Pa,method="svd"))
+    #analysis = exp(analysis)
+    analysis[is.na(analysis)] <- 0
+    analysis <- abs(analysis)
```

Comment:
This seems like a dangerous assumption to build in by default -- the sort of thing that will catch someone years later after a lot of debugging. Many state variables out there are allowed to be negative. If a specific model needs to zero or abs a proposed value that needs to be in the model-specific write.restart. But preferable would be to more intelligently handle non-negative state variables in the filter itself so that negative values are never proposed, otherwise there will be a bias between what the EnKF **thinks** the mean and variance are and what the ensemble mean and variance **actually are**.